File: test/coordinates/test-L1RouteTree.cpp
#include "test/common.h"

#include "marocco/coordinates/L1RouteTree.h"

namespace marocco {

using namespace HMF::Coordinate;

class ASimpleL1Route : public ::testing::Test
{
protected:
	L1Route route{HICANNOnWafer(X(5), Y(5)),
	              Merger0OnHICANN(2),
	              DNCMergerOnHICANN(2),
	              HLineOnHICANN(46),
	              HICANNOnWafer(X(6), Y(5)),
	              HLineOnHICANN(48),
	              VLineOnHICANN(39),
	              SynapseDriverOnHICANN(left, Y(99)),
	              SynapseOnHICANN(SynapseColumnOnHICANN(5), SynapseRowOnHICANN(0))};
};

TEST(L1RouteTree, canBeCopied)
{
	L1Route route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(46)};
	L1RouteTree tree(route);
	L1RouteTree copy(tree);
	ASSERT_EQ(tree, copy);
}

TEST(L1RouteTree, hasHead)
{
	L1Route route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(46)};
	L1RouteTree tree(route);
	ASSERT_EQ(route, tree.head());
}

TEST(L1RouteTree, canHaveTails)
{
	L1Route route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(48)};
	L1RouteTree tree(route);
	EXPECT_FALSE(tree.has_tails());
	tree.add(L1Route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(48), VLineOnHICANN(39)});
	ASSERT_TRUE(tree.has_tails());
}

TEST(L1RouteTree, hasEmptyTailWhenTargetIsReachedEnPassant)
{
	L1Route route{HICANNOnWafer(X(5), Y(5)), Merger0OnHICANN(2), DNCMergerOnHICANN(2),
	              HLineOnHICANN(46)};
	L1Route route_reached_en_passant(L1Route::sequence_type(route.begin(), std::prev(route.end())));

	{
		L1RouteTree tree(route);
		tree.add(route_reached_en_passant);
		EXPECT_EQ(route_reached_en_passant, tree.head());
		EXPECT_TRUE(tree.has_tails());
		auto tails = tree.tails();
		ASSERT_EQ(2, tails.size());
		EXPECT_FALSE(tails[0].get().empty());
		EXPECT_EQ(route.back(), tails[0].get().head().back());
		EXPECT_TRUE(tails[1].get().empty());
	}

	{
		L1RouteTree tree(route_reached_en_passant);
		tree.add(route);
		EXPECT_EQ(route_reached_en_passant, tree.head());
		EXPECT_TRUE(tree.has_tails());
		auto tails = tree.tails();
		ASSERT_EQ(2, tails.size());
		EXPECT_FALSE(tails[0].get().empty());
		EXPECT_EQ(route.back(), tails[0].get().head().back());
		EXPECT_TRUE(tails[1].get().empty());
	}
}

TEST(L1RouteTree, canBeEmpty)
{
	L1Route route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(46)};
	L1RouteTree tree(route);
	EXPECT_FALSE(tree.empty());
	L1RouteTree empty_tree{};
	EXPECT_TRUE(empty_tree.empty());
}

TEST(L1RouteTree, doesNotAcceptDisjunctRoute)
{
	L1Route route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(46)};
	L1RouteTree tree(route);
	ASSERT_ANY_THROW(tree.add(L1Route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(2)}));
	ASSERT_ANY_THROW(tree.add(L1Route{HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(46)}));

	L1RouteTree empty_tree{};
	ASSERT_NO_THROW(empty_tree.add(L1Route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(2)}));
}

TEST(L1RouteTree, doesNotCompareEqualWhenTailsIsSubset)
{
	L1Route head{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(48), VLineOnHICANN(7)};
	L1RouteTree less_tails(head);
	less_tails.add(L1Route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(48), VLineOnHICANN(39)});
	L1RouteTree more_tails = less_tails;
	more_tails.add(L1Route{HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(48), VLineOnHICANN(71)});
	EXPECT_NE(less_tails, more_tails);
	EXPECT_NE(more_tails, less_tails);
}

} // namespace marocco

File: test/coordinates/test-L1RouteTree.py
import itertools
import unittest

from pyhalco_hicann_v2 import *
from pyhalco_common import *
from pymarocco_coordinates import L1Route, L1RouteTree


class L1RouteTreeTest(unittest.TestCase):
    def test_permutations_have_same_result(self):
        routes = []

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(VLineOnHICANN(39))
        route.append(HLineOnHICANN(49))
        route.append(VLineOnHICANN(7))
        routes.append(route)

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(VLineOnHICANN(39))
        routes.append(route)

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        routes.append(route)

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(VLineOnHICANN(7))
        routes.append(route)

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(VLineOnHICANN(71))
        routes.append(route)

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(46))
        routes.append(route)

        route = L1Route()
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(VLineOnHICANN(39))
        route.append(SynapseDriverOnHICANN(left, Y(99)))
        route.append(SynapseOnHICANN(SynapseColumnOnHICANN(5),
                                     SynapseRowOnHICANN(0)))
        routes.append(route)

        for N in range(2, len(routes)):
            for subset in itertools.combinations(routes, N):
                trees = []
                for perm in itertools.permutations(subset, N):
                    tree = L1RouteTree()
                    for route in perm:
                        tree.add(route)
                    if trees:
                        self.assertEqual(trees[0], tree)

        return routes

    def test_routes_fulfilled_en_passant_have_one_empty_tail(self):
        route = L1Route()
        route.append(HICANNOnWafer(X(5), Y(5)), Merger0OnHICANN(2))
        route.append(DNCMergerOnHICANN(2))
        route.append(HLineOnHICANN(46))
        route.append(HICANNOnWafer(X(6), Y(5)), HLineOnHICANN(48))
        route.append(VLineOnHICANN(39))
        route.append(SynapseDriverOnHICANN(left, Y(99)))
        route.append(SynapseOnHICANN(SynapseColumnOnHICANN(5),
                                     SynapseRowOnHICANN(0)))

        hicann = None
        route_ = L1Route()
        for segment in route:
            if isinstance(segment, HICANNOnWafer):
                hicann = segment
                continue
            if hicann is not None:
                route_.append(hicann, segment)
                hicann = None
            else:
                route_.append(segment)

            if route_ == route:
                continue

            tree = L1RouteTree(route)
            tree.add(route_)
            self.assertEqual(route_, tree.head())
            self.assertTrue(tree.has_tails())
            tails = tree.tails()
            self.assertEqual(2, len(tails))
            self.assertEqual(route.back(), tails[0].head().back())
            self.assertTrue(tails[1].empty())

    def test_tree_can_not_be_modified_via_head(self):
        route = L1Route()
        route.append(HICANNOnWafer(X(5), Y(5)), Merger0OnHICANN(2))
        route.append(DNCMergerOnHICANN(2))

        tree = L1RouteTree(route)
        copy = L1RouteTree(tree)
        # tree.head() should return a copy
        tree.head().append(HLineOnHICANN(46))
        self.assertEqual(copy, tree)

    def test_tree_can_not_be_modified_via_tails(self):
        route = L1Route()
        route.append(HICANNOnWafer(X(5), Y(5)), Merger0OnHICANN(2))
        route.append(DNCMergerOnHICANN(2))
        route.append(HLineOnHICANN(46))

        route_ = L1Route()
        route_.append(HICANNOnWafer(X(5), Y(5)), Merger0OnHICANN(2))
        route_.append(DNCMergerOnHICANN(2))

        tree = L1RouteTree(route)
        tree.add(route_)
        copy = L1RouteTree(tree)
        # tree.tails() should return copies
        tails = tree.tails()
        self.assertFalse(tails[0].empty())
        route = L1Route()
        route.append(HICANNOnWafer(X(5), Y(5)), HLineOnHICANN(46))
        route.append(VLineOnHICANN(8))
        tails[0].add(route)
        self.assertEqual(copy, tree)

    def test_supports_pickling(self):
        import pickle
        for route in self.test_permutations_have_same_result():
            d = pickle.dumps(route)
            self.assertEqual(route, pickle.loads(d))


if __name__ == "__main__":
    unittest.main()
